# **Week 10: Large Sample Hypothesis Testing**

```
.------------------------------------.
|   __  ____  ______  _  ___ _____   |
|  |  \/  \ \/ / __ )/ |/ _ \___  |  |
|  | |\/| |\  /|  _ \| | | | | / /   |
|  | |  | |/  \| |_) | | |_| |/ /    |
|  |_|  |_/_/\_\____/|_|\___//_/     |
'------------------------------------'

---

Through the following examples, we will explore the concepts of (large-sample) hypothesis testing (LSHT)
and examine their practical implications.

## **Pre-Configurating the Notebook**

### **Switching to the R Kernel on Colab**

By default, Google Colab uses Python as its programming language. To use R instead, you’ll need to manually switch the kernel by going to **Runtime > Change runtime type**, and selecting R as the kernel. This allows you to run R code in the Colab environment.

However, our notebook is already configured to use R by default. Unless something goes wrong, you shouldn’t need to manually change runtime type.

### **Importing Required Packages**
**Run the following lines of code**:

In [ ]:
#Do not modify

setwd("/content")

# Remove `MXB107-Notebooks` if exists,
if (dir.exists("MXB107-Notebooks")) {
  system("rm -rf MXB107-Notebooks")
}

# Fork the repository
system("git clone https://github.com/edelweiss611428/MXB107-Notebooks.git")

# Change working directory to "MXB107-Notebooks"
setwd("MXB107-Notebooks")

#
invisible(source("R/preConfigurated.R"))

Loading required package: ggplot2

Loading required package: dplyr


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Loading required package: tidyr

Loading required package: stringr

Loading required package: magrittr


Attaching package: ‘magrittr’


The following object is masked from ‘package:tidyr’:

    extract


Loading required package: IRdisplay

Loading required package: png

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘png’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Loading required package: grid

Loading required package: knitr



**Do not modify the following**

In [ ]:
if (!require("testthat")) install.packages("testthat"); library("testthat")

test_that("Test if all packages have been loaded", {

  expect_true(all(c("ggplot2", "tidyr", "dplyr", "stringr", "magrittr", "knitr") %in% loadedNamespaces()))

})

Loading required package: testthat


Attaching package: ‘testthat’


The following objects are masked from ‘package:magrittr’:

    equals, is_less_than, not




Test passed with 1 success 😀.


### **Demonstrating Example**

A phone company advertises that the average battery life of their phones (when continuously watching videos), denoted as $\mu$, is 8 hours.

To verify this claim, an independent random sample of 31 phones was tested.

In [ ]:
smpl_data = c(7.2, 8.53, 8.07, 7.99, 7.79, 7.77, 8.9, 7.64, 7.35, 8.45, 9.14, 7.93, 7.35, 7.52, 7.41, 8.27,
7.55, 7.5, 8.53, 8.37, 8.17, 8.15, 8.02, 7.63,7.64, 8.83, 8.17, 7.41, 7.7, 8.21, 7.13)

[1] 31

**Write down the null and alternative hypotheses for testing whether the company’s claim is correct.**


Since we are testing whether or not there is evidence *against* the company’s claim that the average battery life is 8 hours, the alternative hypothesis should challenge this claim.  

Because we are not concerned if the battery lasts longer than 8 hours (that would be favorable to consumers), we only test if it is **less** than 8 hours.  

$$
\begin{align}
H_0: \mu &= 8 \\
H_1: \mu &< 8
\end{align}
$$

This is a **left-tailed test** (or often simply referred to as one-tailed or one-sided test) of the mean.

**Approximate the sampling distribution of the sample mean under the null hypothesis.**

Since our sample size is larger than 30, it is reasonable to assume the sample mean approximately has a sampling distribution that is normal.  We do not know the population variance of battery life, but we can estimate it from the sample.

In [ ]:
var(smpl_data)
sd(smpl_data)

length(smpl_data)

mean(smpl_data)


[1] 0.2673452

[1] 0.5170543

[1] 31

[1] 7.945806

$$\bar{x} \sim \mathcal{N}\Big(7.95,\frac{0.267}{31}\Big)$$

**Define the z-test statistic for testing the null hypothesis**


$$
Z = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}} \approx \frac{7.95 - 8}{\frac{0.517}{\sqrt{31}}} \approx -0.538
$$


**Now we want to compute the p-value**

Since we have a left-tailed test, the p-value is given by
$$
P(Z < -0.538 | H_0 \mbox{ is true})
$$

This is the probability of getting our observed test statistic value or more extreme (in the direction of the alternative) given that the null is true.  If this probability is small, it would provide evidence against the null in favour of the alternative.

Since $Z$ is approximately standard normal, we can compute the p-value using the cdf of the standard normal distribution (we need R to help us do that).  The full set of R code to conduct the hypothesis test is below.  

In [ ]:
xbar = mean(smpl_data)
s = sd(smpl_data)
n = 31
# compute the test statistic
Z = (xbar-8)/(s/sqrt(n))
Z

p = pnorm(Z)
p


[1] -0.5835691

[1] 0.2797551

Here we see the p-value is quite large, around 0.28. Thus it was reasonably likely to get the sample we got (or more extreme) if the null was true.  Thus there is insufficient evidence against the null.  That is, from this data, there is no evidence against the assumption that the mean battery life is 8 hours.

## **Try It Yourself**




The following questions are based on the `epa_data` dataset. While you are expected to use R to compute the answers, the underlying concepts are identical to those in pen-and-paper hypothesis testing calculations.


#### **Question 1**

Is there evidence of a difference in the average city mileage between cars manufactured in 2015 and 2020?  Interpret the results for a non-statistician stakeholder.

In [ ]:
epa_data = read.csv("./datasets/epa_data.csv")
epa_data %>% str()

'data.frame':	13569 obs. of  9 variables:
 $ city : int  16 15 16 19 19 19 19 19 19 19 ...
 $ hwy  : int  24 22 22 27 29 24 26 27 29 24 ...
 $ cyl  : int  8 8 8 4 4 4 4 4 4 4 ...
 $ disp : num  5 5 5 2 2 2.4 2.4 2 2 2.4 ...
 $ drive: chr  "Rear-Wheel Drive" "Rear-Wheel Drive" "Rear-Wheel Drive" "Rear-Wheel Drive" ...
 $ make : chr  "Jaguar" "Jaguar" "Jaguar" "Pontiac" ...
 $ model: chr  "XK" "XK" "XK Convertible" "Solstice" ...
 $ trans: chr  "Automatic" "Automatic" "Automatic" "Automatic" ...
 $ year : int  2010 2010 2010 2010 2010 2010 2010 2010 2010 2010 ...


<details>
<summary>▶️ Click to show the solution</summary>

```r
# define mu_2015 and mu_2020 as the true mean city mileage for cars manufactured in 2015 and 2020, respectively
# the null and alternative hypotheses are

#H0: mu_2020 = mu_2015
#H1: mu_2020 != mu_2015

# pull out relevant data
mileage_2015 = epa_data$city[epa_data$year == 2015]
mileage_2020 = epa_data$city[epa_data$year == 2020]

# check if sample sizes are large enough for normal approximation
length(mileage_2015) > 30
length(mileage_2020) > 30

# both sample sizes greater 30, so all good for the normal approximation


# compute the test statistic
z = (mean(mileage_2020) - mean(mileage_2015))/sqrt(var(mileage_2020)/length(mileage_2020) + var(mileage_2015)/length(mileage_2015))
z

# since this is a two-sided test, we need to compute P(|Z| > |z|), which we can do as follows (since the normal distribution is symmetric)
p = 2*pnorm(z, lower.tail = FALSE)
p

# at around 0.00015, the p-value is very small.  Hence there is overwhelming evidence against the null, in favour of the alternative
# Thus there is overwhelming evidence that the average city mileage differs between 2015 and 2020
```
</details>

#### **Question 2**

Is there evidence that the proportion of cars produced with automatic transmissions in 2010 is greater than 0.5? Interpret the results for a non-statistician stakeholder.

<details>
<summary>▶️ Click to show the solution</summary>

```r
# Define p as the true proportion of cars produced with automatic transformation in 2010
# the null and alternative are given by

#H0: p = 0.5
#H1: p > 0.5

# pull out relevant data

transmission_2010 = epa_data$trans[epa_data$year == 2010]
auto_2010 = transmission_2010 == "Automatic"

n = length(auto_2010)
n
p_obs = mean(auto_2010)
p_obs

# check if normal approximation is ok
n*0.5 > 10

# both np > 10 and n(1-p) > 10 so we are ok for the normal approximation

# compute the test statistic
z = (p_obs - 0.5)/sqrt(0.5*0.5/n)
z

# compute the p-value for one sided test (right tailed)
pval = pnorm(z, lower.tail = FALSE)
pval

# The p-value is extremely small, thus there is overwhelming evidence against the null
# Thus there is overwhelming evidence that the proportion of automatics produced in 2010 exceeds 0.5
```
</details>

#### **Question 3**

Is there evidence that the proportion of cars produced with manual transmissions for the years 2011 and 2012 has decreased? Interpret the results for a non-statistician stakeholder.

<details>
<summary>▶️ Click to show the solution</summary>

```r
# Define p_2011 and p_2012 as the true proportion of 2011 and 2012 manual transmission cars, respectively

#H0: p_2012 = p_2011
#H1: p_2011 > p_2012



# pull out observed proportions and other statistics
transmission_2011 = epa_data$trans[epa_data$year == 2011]
transmission_2012 = epa_data$trans[epa_data$year == 2012]

manual_2011 = transmission_2011 == "Manual"
manual_2012 = transmission_2012 == "Manual"

n1 = length(manual_2011)
n2 = length(manual_2012)
n1
n2

p_obs1 = mean(manual_2011)
p_obs2 = mean(manual_2012)
p_obs1
p_obs2

# check conditions for normal approximation
p_pool = (sum(manual_2011) + sum(manual_2012))/(n1 + n2)
p_pool

n1*p_pool > 10
n1*(1-p_pool) > 10

n2*p_pool > 10
n2*(1-p_pool) > 10

# all the tests pass, ok for the normal approximation

# compute the test statistic
z = (p_obs1 - p_obs2)/sqrt(p_pool*(1 - p_pool)*(1/n1 + 1/n2))
z

# compute the pvalue for the one sided test
pval = 1-pnorm(z)
pval

# At around 0.36 the p-value is large, hence there is insuficient evidence against the null
# Thus there is no evidence that the proportion of cars produced with manual transmissions for the years 2011 and 2012 has decreased


```
</details>

## **Workshop Questions**




### **Question 1**

The following questions are based on the `episodes` dataset. While you are expected to use R to compute the answers, the underlying concepts are identical to those in pen-and-paper hypothesis testing calculations.

In [ ]:
episodes = read.csv("./datasets/episodes.csv")
episodes %>% str()

#### **Question 1.1**

Is there evidence that the average IMDB rating of Star Trek: The Original Series episodes is lower than 7.7? Interpret the results for a non-statistician stakeholder.


<details>
<summary>▶️ Click to show the solution</summary>

Solutions will be released at the end of the week!

</details>


#### **Question 1.2**


Is there evidence that the mean IMDB rating of episodes of Star Trek: The Original Series differs from the mean rating of Star Trek: The Next Generation? Interpret the results for a non-statistician stakeholder.


<details>
<summary>▶️ Click to show the solution</summary>

Solutions will be released at the end of the week!

</details>


#### **Question 1.3**

Is there evidence that the proportion of Star Trek: The Next Generation episodes that pass the Bechdel-Wallace Test is different from 0.4? Interpret the results for a non-statistician stakeholder.


**Note**: While these series have ended and you technically have the full “population” of some values (e.g., results of the Bechdel-Wallace test), we still ask you to test whether or not the proportion of episodes that pass the test is equal to 0.4. This may seem counter-intuitive, but you can think of it as follows:

- The episode test results are treated as realisations from an unknown probability distribution $f$ (here, Bernoulli(p)).
- Although the episodes are released, we are interested in the underlying process that generates these values. This includes not-yet-released episodes or hypothetical similar episodes. Simply examining the “complete” population of Bechdel test results is not sufficient; instead, we rely on a statistical model to quantify uncertainty.


<details>
<summary>▶️ Click to show the solution</summary>

Solutions will be released at the end of the week!

</details>


#### **Question 1.4**


Is there evidence that the proportion of episodes that pass the Bechdel-Wallace Test differs between Star Trek: The Next Generation and Star Trek: Voyager? Interpret the results for a non-statistician stakeholder.


<details>
<summary>▶️ Click to show the solution</summary>

Solutions will be released at the end of the week!

</details>
